src/dmd/ctorflow.d

```diff
@@ -105,53 +105,118 @@ struct CtorFlow
  * Returns:
- *      false means either fieldInit or fi skips initialization
+ *      false means one of the paths skips construction
  */
-bool mergeFieldInitX(ref CSX fieldInit, const CSX fi)
+bool mergeCallSuper(ref CSX a, const CSX b) pure nothrow
 {
-    if (fi != fieldInit)
+    // This does a primitive flow analysis to support the restrictions
+    // regarding when and how constructors can appear.
+    // It merges the results of two paths.
+    // The two paths are `a` and `b`; the result is merged into `a`.
+    bool ok = true;
+    if (b != a)
     {
+        // Have ALL branches called a constructor?
+        int aAll = (a & (CSX.this_ctor | CSX.super_ctor)) != 0;
+        int bAll = (b & (CSX.this_ctor | CSX.super_ctor)) != 0;
+        // Have ANY branches called a constructor?
+        bool aAny = (a & CSX.any_ctor) != 0;
+        bool bAny = (b & CSX.any_ctor) != 0;
         // Have any branches returned?
-        bool aRet = (fi & CSX.return_) != 0;
-        bool bRet = (fieldInit & CSX.return_) != 0;
+        bool aRet = (a & CSX.return_) != 0;
+        bool bRet = (b & CSX.return_) != 0;
         // Have any branches halted?
-        bool aHalt = (fi & CSX.halt) != 0;
-        bool bHalt = (fieldInit & CSX.halt) != 0;
-        bool ok;
+        bool aHalt = (a & CSX.halt) != 0;
+        bool bHalt = (b & CSX.halt) != 0;
         if (aHalt && bHalt)
         {
-            ok = true;
-            fieldInit = CSX.halt;
+            a = CSX.halt;
         }
-        else if (!aHalt && aRet)
+        else if ((!bHalt && bRet && !bAny && aAny) || (!aHalt && aRet && !aAny && bAny))
+        {
+            // If one has returned without a constructor call, there must not
+            // be ctor calls in the other.
+            ok = false;
+        }
+        else if (bHalt || bRet && bAll)
+        {
+            // If one branch has called a ctor and then exited, anything the
+            // other branch has done is OK (except returning without a
+            // ctor call, but we already checked that).
+            a |= b & (CSX.any_ctor | CSX.label);
+        }
+        else if (aHalt || aRet && aAll)
+        {
+            a = cast(CSX)(b | (a & (CSX.any_ctor | CSX.label)));
+        }
+        else
+        {
+            // Both branches must have called ctors, or both not.
+            ok = (aAll == bAll);
+            // If one returned without a ctor, we must remember that
+            // (Don't bother if we've already found an error)
+            if (ok && bRet && !bAny)
+                a |= CSX.return_;
+            a |= b & (CSX.any_ctor | CSX.label);
+        }
+    }
+    return ok;
+}
+
+
+/****************************************
+ * Merge `b` flow analysis results into `a`.
+ * Params:
+ *      a = the path to merge `b` into
+ *      b = the other path
+ * Returns:
+ *      false means either `a` or `b` skips initialization
+ */
+bool mergeFieldInit(ref CSX a, const CSX b) pure nothrow
+{
+    if (b != a)
```

Comment:
here we go again

Reply:
Well, I beg that that code was all copy-pasta from the original.